server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java

```diff
@@ -271,6 +276,70 @@ public void testDeleteDanglingIndicesRequiresDataLossFlagToBeTrue() throws Excep
         assertThat(e.getMessage(), containsString("accept_data_loss must be set to true"));
     }
 
+    /**
+     * Check that import-and-delete of a dangling index doesn't have a race condition that bypasses the graveyard and permits a re-import.
+     */
+    public void testDanglingIndicesImportedAndDeletedCannotBeReimported() throws Exception {
+
+        final Settings settings = buildSettings(1, true);
+        internalCluster().startNodes(3, settings);
+
+        final String stoppedNodeName = createDanglingIndices(INDEX_NAME, OTHER_INDEX_NAME);
+        final String danglingIndexUUID = findDanglingIndexForNode(stoppedNodeName, INDEX_NAME);
+
+        final AtomicBoolean isImporting = new AtomicBoolean();
+        final Thread[] importThreads = new Thread[2];
+        for (int i = 0; i < importThreads.length; i++) {
+            importThreads[i] = new Thread(() -> {
+                //noinspection StatementWithEmptyBody
+                while (isImporting.get() == false) {
```

Comment:
nit: I prefer a latch just to not burn cycles on CI unnecessarily. Also, this avoids the very unlikely edge case of the main thread passing through 10 seconds of trying to delete the index either before the thread gets here or in one of the loop iterations.

Reply:
:+1: see https://github.com/elastic/elasticsearch/pull/67687/commits/22859605916e82a54148a23acde11f60b5dffc54